include/cantera/kinetics/StoichManager.h

```diff
@@ -474,6 +691,31 @@ class StoichManagerN
         m_stoichCoeffs.reserve(nCoeffs);
         m_stoichCoeffs.setFromTriplets(m_coeffList.begin(), m_coeffList.end());
 
+        // Set up outer/inner indices for mapped jacobian output
+        Eigen::SparseMatrix<double> tmp = m_stoichCoeffs.transpose();
+        m_outerIndices.resize(nSpc + 1); // number of columns + 1
+        for (int i = 0; i < tmp.outerSize() + 1; i++) {
+            m_outerIndices[i] = tmp.outerIndexPtr()[i];
+        }
+        m_innerIndices.resize(nCoeffs);
+        for (size_t n = 0; n < nCoeffs; n++) {
+            m_innerIndices[n] = tmp.innerIndexPtr()[n];
+        }
+        m_values.resize(nCoeffs, 0.);
+
+        // Set up index pairs for jacobians
+        std::vector<std::pair<int, int>> indices;
+        for (int i = 0; i < tmp.outerSize(); i++) {
+            for (Eigen::SparseMatrix<double>::InnerIterator it(tmp, i); it; ++it) {
+                indices.emplace_back(it.row(), it.col());
+            }
+        }
```

Comment:
Assuming I've followed the logic correctly, I think the following would substantially simplify the implementation of the `updateSetup` methods:
```suggestion
        std::map<std::pair<int, int>, size_t> indices;
        size_t n = 0;
        for (int i = 0; i < tmp.outerSize(); i++) {
            for (Eigen::SparseMatrix<double>::InnerIterator it(tmp, i); it; ++it) {
                indices[{it.row(), it.col()}] = n++;
            }
        }
```

Reply:
Could you elaborate? Wouldn't iterating over a  `map` be slower than a classic loop over a `vector`, where `n` is increased automatically? (see `Cx:updateSetup`, which I renamed to `Cx:resizeCoeffs` for consistency). 

As an aside, buffering these indices is key to being able to directly write into an `Eigen::SparseMatrix` while avoiding costly index look-ups. I believe @anthony-walker may have adopted this in #1010, as we discussed this a while back: these offsets can also be used for any composite Jacobian that is assembled from `Kinetics`.

Reply:
With this map structure, though, you don't need to iterate and check each of the elements -- you just look up the element using the (reaction index, species index) tuple as the key. I went ahead and implemented this to demonstrate -- feel free to cherry-pick https://github.com/speth/cantera/commit/f1c1e686bcb09763c5efc8f2a19ddbee9bd350a9.

It's both a lot less code, and faster. Looking up an element in the map is O(log n), while the linear search is O(n); using an `unordered_map` would actually cut this down to O(1), but I didn't think it was worth writing a custom hash function for.

Reply:
Cherry-picked the suggested update.